crates/router/src/services/authentication.rs

```diff
@@ -704,6 +711,75 @@ where
     decode_jwt(&token, state).await
 }
 
+#[async_trait]
+impl<A> AuthenticateAndFetch<AuthenticationDataOrg, A> for JWTAuth
+where
+    A: AppStateInfo + Sync,
+{
+    async fn authenticate_and_fetch(
+        &self,
+        request_headers: &HeaderMap,
+        state: &A,
+    ) -> RouterResult<(AuthenticationDataOrg, AuthenticationType)> {
+        let payload = parse_jwt_payload::<A, AuthToken>(request_headers, state).await?;
+        if payload.check_in_blacklist(state).await? {
+            return Err(errors::ApiErrorResponse::InvalidJwtToken.into());
+        }
+
+        let permissions = authorization::get_permissions(state, &payload).await?;
+        authorization::check_authorization(&self.0, &permissions)?;
+        let org_id = payload.org_id;
+
+        let key_store = state
+            .store()
+            .get_merchant_key_store_by_merchant_id(
+                &payload.merchant_id,
+                &state.store().get_master_key().to_vec().into(),
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::InvalidJwtToken)
+            .attach_printable("Failed to fetch merchant key store for the merchant id")?;
+        //get merchant ids for above org_id
+        let merchant = state
+            .store()
+            .find_merchant_account_by_merchant_id(&payload.merchant_id, &key_store)
+            .await
+            .change_context(errors::ApiErrorResponse::InvalidJwtToken)?;
+        let organization = state
+            .store()
+            .list_merchant_accounts_by_organization_id(&org_id)
+            .await
+            .change_context(errors::ApiErrorResponse::InvalidJwtToken)?;
+        let org_merchant_ids: Vec<String> = permissions
+            .iter()
+            .filter_map(|permission| match permission {
+                Permission::OrgAnalytics => Some(
+                    organization
+                        .iter()
+                        .map(|org| org.merchant_id.clone())
+                        .collect(),
+                ),
+                _ => None,
```

Comment:
```suggestion
                _ => vec![&merchant.merchant_id.clone()],
```

if the user doesn't have org admin access return only a single merchant id that he is belongs to.

Reply:
it need to be none or in the every iteration it will try to push the merchant_id when the permission is not matching